### ISSUE TITLE
Rework argument comparison

### ISSUE THREAD

**Issue by rusty-snake** [libseccomp-rs/libseccomp-rs#41]
#### How does argument comparison work in libseccomp?

You can either act on a syscall unconditional (no argument comparison) or conditional (i.e. based on the argument of the syscall. And no you can not dereference pointers.).

If you use argument comparison, you specify rules which compare the nth-value of a syscall to a datum value using a `scmp_compare` operator.

These rules are logically AND-ed:

> If syscall argument comparisons are included in the filter rule, all of the comparisons must be  true  for  the  rule  to match.

But even more important, you can only have one comparisons for an argument:

> When adding syscall argument comparisons to the filter it is important to remember that while it is possible to have multiple comparisons in a single rule, you can only compare each argument once in a single rule.  In other  words,  you  can not have multiple comparisons of the 3rd syscall argument in a single rule.

If you try to add multiple comparisons for one argument, `add_rule` will return `Err(_)`. That's perfectly fine. However from an user PoV it woulb be great if the API would not all this.

#### Ideas

1. Adopt libseccomp-golang like `add_rule*` methods:
   ```rust
   fn add_rule(&mut self, syscall: i32, action ScmpAction) -> Result<()>;
   fn add_rule_conditional(&mut self, syscall: i32, action ScmpAction, comparators: &[ScmpArgCompare]) -> Result<()>;
   fn add_rule_conditional_exact(&mut self, syscall: i32, action ScmpAction, comparators: &[ScmpArgCompare]) -> Result<()>;
   fn add_rule_exact(&mut self, syscall: i32, action ScmpAction) -> Result<()>;
   ```
2. Add an own `ArgumentComparators` type:
   ```rust
   // Same ABI: `&[libseccomp-sys::scmp_arg_cmp] as *const libseccomp-sys::scmp_arg_cmp` <=> `const struct scmp_arg_cmp *`
   struct ArgComp([libseccomp-sys::scmp_arg_cmp; 6]);
   impl ArgComp {
       pub fn get(n: usize) -> ScmpArgCompare;
       pub fn set(n: usize, op: ScmpCompareOp, datum: u64);
       pub fn set_noreplace(n: usize, op: ScmpCompareOp, datum: u64);
       pub fn from_iter(i: impl IntoIterator<Item = (ScmpCompareOp, u64)>) -> Result<Self>`
       pub fn from_iter(i: impl IntoIterator<Item = ScmpArgCompare) -> Result<Self>`
   }
   ```
   _As you see this is an unfinish idea._
3. (Ab)use rust's operators:
   ```rust
   ctx.add_rule(ScmpAction::Allow, 1000, &[ARG0 == 5, ARG1 != 2, ARG2 >= 64, ARG3 & 0x0f0 == 1234])
   ```
   This is rust, not python so we need `cmp![ARG0 == …`.

Basically I want to simplify
```rust
ctx.add_rule(ScmpAction::Allow, 1000, Some(&[ScmpArgCompare::new(0, ScmpCompareOp::Equal, 1), ScmpArgCompare::new(2, ScmpCompareOp::NotEqual, 32)]))?;
```
to something like
```rust
ctx.add_rule(ScmpAction::Allow, 1000, ArgCmp(arg_0_eq(1), None, arg_2_ne(32)))?;
```

Obvious this needs some time, thinking, input, ...

**Comment by rusty-snake** [comment]
My current idea is to go with 1. (i.e. libseccomp-golang add_rule\* but with flipped `syscall`/`action` parameters) and 3. an macro based argument compare DSL.

Old:
```rust
ctx.add_rule(ScmpAction::Allow, 1000, Some(&[ScmpArgCompare::new(0, ScmpCompareOp::Equal, 1), ScmpArgCompare::new(2, ScmpCompareOp::NotEqual, 32), ScmpArgCompare::new(2, ScmpCompareOp::MaskedEqual(0x0f0), 22)]))?;
```

New:
```rust
ctx.add_rule(ScmpAction::Allow, 1000, &[scmp_cmp!($arg0 == 1), scmp_cmp!($arg2 != 32), scmp_cmp!($arg2 & 0x0f0 == 22)])?;
```

My dream is 
```rust
ctx.add_rule(ScmpAction::Allow, 1000, scmp_cmp![$arg0 == 1, $arg2 != 32, $arg2 & 0x0f0 == 22])?;
```
but I could not find a way to make it work (without proc_macro hell).
